File: 02-Web开发/05-流量治理与限流/01-GCRA与漏桶计量/python/gcra.py
"""GCRA 与漏桶计量。

权威依据：
- ITU-T I.371《Traffic control and congestion control in B-ISDN》给出的两种等价描述
  （虚拟调度 virtual scheduling / 连续状态漏桶 continuous-state leaky bucket），
  原文见 Wikipedia《Generic cell rate algorithm》对该建议书的引述。
- Brandur Leach《Rate Limiting, Cells, and GCRA》(brandur.org/rate-limiting)。

三种计量器在本模块中并列实现，便于逐事件对拍：
- FixedWindow：时间桶（按窗口网格对齐）。
- Gcra：虚拟调度形态，状态只有一个标量 TAT。
- ContinuousLeakyBucket：连续状态漏桶，状态是 (X, LCT)。
"""

from __future__ import annotations


class FixedWindow:
    """时间桶（time bucketed）：窗口网格对齐，计数在窗口边界清零。"""

    def __init__(self, limit: int, window: float) -> None:
        self.limit = limit
        self.window = window
        self.window_start = None
        self.count = 0

    def arrive(self, ta: float, cost: int = 1) -> bool:
        if self.window_start is None or ta - self.window_start >= self.window:
            self.window_start = (ta // self.window) * self.window
            self.count = 0
        if self.count + cost <= self.limit:
            self.count += cost
            return True
        return False


class Gcra:
    """虚拟调度形态的 GCRA。

    T    = 1/rate，发射间隔（秒）。
    tau  = 容限（秒），即「最多可以比理论到达时间提前多久」。

    buffer_mode:
      "itu"     —— ITU-T I.371 / Wikipedia 口径：next_allowed = TAT - tau。
      "brandur" —— Brandur 文章口径：扣减 (tau + T)，等于多发一张令牌。
    """

    def __init__(self, rate: float, tau: float, strict: bool = False,
                 buffer_mode: str = "itu") -> None:
        assert rate > 0, "rate 必须为正"
        assert tau >= 0, "容限不能为负"
        self.T = 1.0 / rate
        self.tau = tau
        self.strict = strict          # True 用规范原文的 ta > next_allowed
        self.buffer_mode = buffer_mode
        self.tat = None               # 理论到达时间；None 表示尚未收到任何请求

    @property
    def buffer(self) -> float:
        return self.tau if self.buffer_mode == "itu" else self.tau + self.T

    def next_allowed(self):
        if self.tat is None:
            return None
        return self.tat - self.buffer

    def arrive(self, ta: float, cost: int = 1):
        """一次到达。返回 (conforming, retry_after)。非一致时状态不变。"""
        if self.tat is None:
            self.tat = ta + cost * self.T
            return True, 0.0
        na = self.next_allowed()
        ok = (ta > na) if self.strict else (ta >= na)
        if not ok:
            return False, na - ta
        self.tat = max(self.tat, ta) + cost * self.T
        return True, 0.0

    def burst_capacity(self) -> int:
        """同一瞬间最多能连续通过多少次（cost=1）。"""
        g = Gcra(1.0 / self.T, self.tau, self.strict, self.buffer_mode)
        n = 0
        while True:
            ok, _ = g.arrive(0.0)
            if not ok:
                return n
            n += 1


class ContinuousLeakyBucket:
    """连续状态漏桶（ITU-T I.371）：桶内以 1 单位/单位时间漏水，一致信元加 T。

    桶容量上界是 (T + tau)；判据是 X' <= tau。
    """

    def __init__(self, rate: float, tau_limit: float) -> None:
        assert rate > 0
        assert tau_limit >= 0
        self.T = 1.0 / rate
        self.tau = tau_limit
        self.X = 0.0       # 桶内水量（单位是时间）
        self.lct = None    # 上一次一致信元的到达时刻

    def arrive(self, ta: float, cost: int = 1):
        if self.lct is None:
            self.lct = ta
            self.X = cost * self.T
            return True, 0.0
        xp = self.X - (ta - self.lct)   # 漏水
        if xp < 0.0:
            xp = 0.0
        if xp > self.tau:
            return False, xp - self.tau  # 漏水速率为 1，故差值即还需等待的时间
        self.lct = ta
        self.X = xp + cost * self.T
        return True, 0.0


def measure_throughput(limiter, rate: float, horizon: float = 10.0) -> int:
    """以远快于限额的速度持续冲击，统计 horizon 秒内通过的次数。"""
    n = 0
    t = 0.0
    step = 1.0 / (rate * 1000.0) if rate > 0 else 0.001
    while t < horizon:
        ok, _ = limiter.arrive(t)
        if ok:
            n += 1
        t += step
    return n

File: 02-Web开发/05-流量治理与限流/01-GCRA与漏桶计量/go/gcra.go
// Package gcra 实现 GCRA 与连续状态漏桶的等价对照。
//
// 依据 ITU-T I.371 的两种等价描述（经 Wikipedia《Generic cell rate algorithm》转述）：
//   - 虚拟调度：维护理论到达时间 TAT，一致判据 ta >= TAT - tau；
//     一致时 TAT = max(TAT, ta) + T，非一致时 TAT 不变。
//   - 连续状态漏桶：桶内以 1 单位/单位时间漏水，一致信元加 T，判据 X' <= tau。
package gcra

// FixedWindow 时间桶：按窗口网格对齐，计数在窗口边界清零。
type FixedWindow struct {
	Limit       int
	Window      float64
	WindowStart float64
	Started     bool
	Count       int
}

// Arrive 返回本次是否放行。
func (f *FixedWindow) Arrive(ta float64) bool {
	if !f.Started || ta-f.WindowStart >= f.Window {
		f.WindowStart = float64(int(ta/f.Window)) * f.Window
		f.Started = true
		f.Count = 0
	}
	if f.Count+1 <= f.Limit {
		f.Count++
		return true
	}
	return false
}

// Gcra 虚拟调度形态。BufferMode 为 "itu"（扣 tau）或 "brandur"（扣 tau+T）。
type Gcra struct {
	T          float64
	Tau        float64
	Strict     bool
	BufferMode string
	Tat        float64
	HasTat     bool
}

// NewGcra 由速率与容限构造。
func NewGcra(rate, tau float64) *Gcra {
	if rate <= 0 {
		panic("rate must be positive")
	}
	return &Gcra{T: 1.0 / rate, Tau: tau, BufferMode: "itu"}
}

func (g *Gcra) buffer() float64 {
	if g.BufferMode == "brandur" {
		return g.Tau + g.T
	}
	return g.Tau
}

// NextAllowed 返回下一次允许的时刻；尚无状态时返回 ok=false。
func (g *Gcra) NextAllowed() (float64, bool) {
	if !g.HasTat {
		return 0, false
	}
	return g.Tat - g.buffer(), true
}

// Arrive 处理一次到达，返回 (是否一致, 还需等待秒数)。
func (g *Gcra) Arrive(ta float64, cost int) (bool, float64) {
	if !g.HasTat {
		g.Tat = ta + float64(cost)*g.T
		g.HasTat = true
		return true, 0
	}
	na, _ := g.NextAllowed()
	ok := ta >= na
	if g.Strict {
		ok = ta > na
	}
	if !ok {
		return false, na - ta
	}
	g.Tat = max(g.Tat, ta) + float64(cost)*g.T
	return true, 0
}

// BurstCapacity 同一瞬间最多连续通过多少次。
func (g *Gcra) BurstCapacity() int {
	probe := NewGcra(1.0/g.T, g.Tau)
	probe.Strict = g.Strict
	probe.BufferMode = g.BufferMode
	n := 0
	for {
		ok, _ := probe.Arrive(0.0, 1)
		if !ok {
			return n
		}
		n++
	}
}

// ContinuousLeakyBucket 连续状态漏桶，桶容量上界为 T + tau。
type ContinuousLeakyBucket struct {
	T    float64
	Tau  float64
	X    float64
	Lct  float64
	Init bool
}

// NewContinuousLeakyBucket 由速率与容限构造。
func NewContinuousLeakyBucket(rate, tau float64) *ContinuousLeakyBucket {
	return &ContinuousLeakyBucket{T: 1.0 / rate, Tau: tau}
}

// Arrive 处理一次到达，返回 (是否一致, 还需等待秒数)。
func (b *ContinuousLeakyBucket) Arrive(ta float64, cost int) (bool, float64) {
	if !b.Init {
		b.Init = true
		b.Lct = ta
		b.X = float64(cost) * b.T
		return true, 0
	}
	xp := b.X - (ta - b.Lct)
	if xp < 0 {
		xp = 0
	}
	if xp > b.Tau {
		return false, xp - b.Tau
	}
	b.Lct = ta
	b.X = xp + float64(cost)*b.T
	return true, 0
}

func max(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}
